Implement email verification features

Malwina9086 opened the pull request:
- Added 'verify' and 'verificationToken' fields to User model in user.schema.js.
- Implemented verification logic in user.js (controller):
  - Created 'verifyUser' function to update user verification status.
  - Implemented 'sendVerificationEmail' function to send email with verification link.
  - Added 'resendVerificationEmail' function for resending verification emails.
- Updated auth.js (routes/api):
  - Added POST /verify endpoint for resending verification emails.
  - Handled validation, user lookup, and response messages.

Malwina9086 committed:
Implement email verification features

- Added 'verify' and 'verificationToken' fields to User model in user.schema.js.
- Implemented verification logic in user.js controller:
  - Created 'verifyUser' function to update user verification status.
  - Implemented 'sendVerificationEmail' function to send email with verification link.
  - Added 'resendVerificationEmail' function for resending verification emails.
- Updated auth.js routes:
  - Added POST /verify endpoint for resending verification emails.
  - Handled validation, user lookup, and response messages.

KamilMr commented on controller/user.js:
```diff
@@ -128,4 +130,92 @@ const updateAvatarUser = async (file, userId) => {
+
+    return { message: "Verification successful" };
+  } catch (error) {
+    console.error("Verify User Error:", error.message);
```
Ta wiadomość jest dla Ciebie więc lepiej sobie ją poszerzyć abyś od razu miała pełen zestaw informacji na temat błędu czyli zamiast error.message uzyc error

KamilMr commented on controller/user.js:
```diff
@@ -128,4 +130,92 @@ const updateAvatarUser = async (file, userId) => {
+      from: process.env.EMAIL_USER,
+      to: email,
+      subject: "Verification Email",
+      text: `Click the following link to verify your email: http://your_api_base_url/users/verify/${verificationToken}`,
```
```suggestion
      text: `Click the following link to verify your email: http://localhost:3000/users/verify/${verificationToken}`,
```
Tu gdzie jest ta zmiana to musisz podać twój prawdziwy port. Normalnie to to byłby cały adres do twojego serwera.

KamilMr replied:
Rozumiem że jeszcze tego nie testowałaś, prawda?

KamilMr commented on controller/user.js:
```diff
@@ -128,4 +130,92 @@ const updateAvatarUser = async (file, userId) => {
+
+const sendVerificationEmail = async (email, verificationToken) => {
+  try {
+    const transporter = nodemailer.createTransport({
```
Ten moduł do wysyłania wiadomości przenieś do innego pliku. A następnie użyj go w tym miejscu. Dzięki takiemu podejściu będziesz mogła używać go nie tylko dla użytkownika ale też przy innych sytuacjach. Taka funkcja wysyłająca e-mail powinna mieć konfigurację przy sobie przyjmować parametry i wysłać email.

sendVerificationEmail powinno przygotować parametry czyli do kogo jaka treść itp.

KamilMr commented on routes/api/auth.js:
```diff
@@ -29,6 +29,28 @@ router.post("/login", async (req, res, next) => {
   }
 });
 
+router.post("/verify", async (req, res, next) => {
```
Ta ścieżka służy do ponownego wysłania wiadomości brakuje ci ścieżki która weryfikuje token.

KamilMr reviewed (changes requested):
Hej,
Kilka komentarzy w kodzie brakuje jeszcze żeby kiedy użytkownik się rejestruje i tworzy swoje konto żeby wysłać Wtedy pierwszy raz w wiadomość email do niego.

Oprócz tego dobrze jest abyś przetestowała sobie to, to znaczy faktycznie wysłała wiadomość odebrała ją na swojej poczcie a następnie żeby cię ona przeniosła do Twojego konta.

Malwina9086 committed:
Organize code and enhance email verification

- Improved comments in the previous PR.
- Moved email sending to `Services/emailServices.js`.
- Reorganized project structure by relocating `schemas` to the `Services` folder.
- Added token verification to the email resend route.

Testing:
- Confirm clarity of updated comments.
- Ensure email sending still works post-refactor.
- Validate `schemas` relocation to `Services` folder.
- Test email resend route with token verification.

KamilMr commented on services/emailService.js:
```diff
@@ -0,0 +1,17 @@
+  from: "malwina@gmail.com",
+  subject: "Sending with SendGrid is Fun",
+  text: "and easy to do anywhere, even with Node.js",
+  html: "<strong>and easy to do anywhere, even with Node.js</strong>",
```
Jeszcze dodaj aby ten html był generowany. Bo teraz Ci to wysyła email, ale nie wysyła linku z tokenem. Bo zobacz, ten plik powinien być opakowany w funkcje która przyjmie argumenty do wysyłki. Teraz będzie wysyłał statycznie generowaną wiadomość.

KamilMr commented on services/emailService.js:
```diff
@@ -0,0 +1,17 @@
+const sgMail = require("@sendgrid/mail");
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+const msg = {
+  to: "malwina@gmail.com",
```
To te powinno być generowane. W testach może być stały email, ale normalnie to to ma przyjsc na rozne email ktore sa podawane.

KamilMr reviewed (changes requested):
Hej, małe naprawy jeszcze. Myslę że janso napisane w pr, jak nie to daj znać to uszczegółowie. :) Powodzenia

Malwina9086 committed:
Improve email sending functionality

- Introduce sendVerificationEmail function to handle email sending with recipient email and verification token as parameters.
- Move SendGrid API key configuration to the main function for better modularity.
- Replace static email content (recipient email, subject, etc.) with dynamic values fetched from environment variables.
- Utilize process.env.SENDGRID_EMAIL_USER and process.env.SENDGRID_API_KEY as values for sender and recipient, enhancing flexibility and security.

KamilMr reviewed (changes requested):
Na filmiku komantarz. Dopowiem, że daje teraz że Trzeba dopracować ale jeżeli goni cie czas i potrzebujesz to mieć już merged żeby robić dalsze część zadania to zapisz sobie w kodzie co musisz jeszcze zmienić wyślij mi ponownie ja to zatwierdze. A ewentualne zmiany zrobisz później i podeślesz.

Malwina9086 committed:
Refactor email sending function to be asynchronous

Transformed the original email sending function into an asynchronous function to leverage the 'await' keyword for handling asynchronous operations, such as sending emails using SendGrid. Removed unnecessary 'return' statements and passed 'email' and 'verificationToken' parameters directly to the function instead of relying on process.env values.

KamilMr reviewed (approved):
Teraz wyglada spoko :)
